File: server-dotnet/src/Mdbrain/Features/Sync/SyncContracts.cs
using System.Text.Json.Serialization;

namespace Mdbrain.Features.Sync;

public sealed record HashEntry(
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("hash")] string Hash);

public sealed record SyncChangesRequest(
    [property: JsonPropertyName("notes")] IReadOnlyList<HashEntry>? Notes,
    [property: JsonPropertyName("assets")] IReadOnlyList<HashEntry>? Assets);

public sealed record SyncNoteRequest(
    [property: JsonPropertyName("path")] string? Path,
    [property: JsonPropertyName("content")] string? Content,
    [property: JsonPropertyName("hash")] string? Hash,
    [property: JsonPropertyName("metadata")] Dictionary<string, object?>? Metadata,
    [property: JsonPropertyName("assets")] IReadOnlyList<HashEntry>? Assets,
    [property: JsonPropertyName("linked_notes")] IReadOnlyList<HashEntry>? LinkedNotes);

public sealed record SyncAssetRequest(
    [property: JsonPropertyName("path")] string? Path,
    [property: JsonPropertyName("contentType")] string? ContentType,
    [property: JsonPropertyName("size")] long? Size,
    [property: JsonPropertyName("hash")] string? Hash,
    [property: JsonPropertyName("content")] string? Content);

File: packages/shared/src/sync-contracts.ts
export interface SyncSnapshotEntry {
  id: string;
  hash: string;
}

export interface SyncChangesRequest {
  notes: SyncSnapshotEntry[];
  assets: SyncSnapshotEntry[];
}

export interface SyncChangesResponse {
  success: boolean;
  need_upsert?: {
    notes: SyncSnapshotEntry[];
    assets: SyncSnapshotEntry[];
  };
  deleted_on_server?: {
    notes: SyncSnapshotEntry[];
    assets: SyncSnapshotEntry[];
  };
  error?: string;
}

export interface SyncNoteRequest {
  path: string;
  content: string;
  hash: string;
  metadata?: Record<string, unknown>;
  assets: Array<{ id: string; hash: string }>;
  linked_notes: Array<{ id: string; hash: string }>;
}

export interface SyncAssetRequest {
  path: string;
  contentType: string;
  size: number;
  hash: string;
  content: string;
}

export interface VaultInfoResponse {
  success: boolean;
  vault?: {
    id: string;
    name: string;
    domain?: string;
    createdAt?: string;
  };
  error?: string;
}
